From 9cc015336954e324b2c7eec9dc716e0430fabb72 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Sat, 25 Jan 2025 13:00:30 +0800
Subject: [PATCH 169/288] AOSCOS: MIPS: platform: disable unreliable CSR-based
 temperature reading
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Disable CSR-based temperature reading, found only on Loongson 3A/B4000
processors. We found that on the Excelsior E23L (3B4000) boards, the
temperature reading would erratically pin to 239°C, triggering immediate
thermal shutdown.

We don't yet know about the true cause behind this issue, as both boards
we own are indeed using non-original heatsinks (but they can in fact
sustain >= 6 hours of stress testing with s-tui).

For now, disable this feature so that we can keep those systems working
and Loongson 3A/B4000 can still report their temperatures via the Chip
Temperature register (loongson_chiptemp).

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Link: https://t.me/c/1109254909/736238
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/platform/mips/cpu_hwmon.c | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/drivers/platform/mips/cpu_hwmon.c b/drivers/platform/mips/cpu_hwmon.c
index 2ac2f31090f9..93619227f039 100644
--- a/drivers/platform/mips/cpu_hwmon.c
+++ b/drivers/platform/mips/cpu_hwmon.c
@@ -135,9 +135,22 @@ static int __init loongson_hwmon_init(void)
 {
 	pr_info("Loongson Hwmon Enter...\n");
 
+	/*
+	 * FIXME: Disable CSR-based temperature reading, found only on Loongson
+	 * 3A/B4000 processors. We found that on the Excelsior E23L (3B4000)
+	 * boards, the temperature reading would erratically pin to 239°C,
+	 * triggering immediate thermal shutdown.
+	 *
+	 * We don't yet know about the true cause behind this issue, as both
+	 * boards we own are indeed using non-original heatsinks (but they can
+	 * in fact sustain >= 6 hours of stress testing with s-tui).
+	 *
+	 * For now, disable this feature so that we can keep those systems
+	 * working and Loongson 3A/B4000 can still report their temperatures
+	 * via the Chip Temperature register (loongson_chiptemp).
+	 */
 	if (cpu_has_csr())
-		csr_temp_enable = csr_readl(LOONGSON_CSR_FEATURES) &
-				  LOONGSON_CSRF_TEMP;
+		csr_temp_enable = 0;
 
 	if (!csr_temp_enable && !loongson_chiptemp[0])
 		return -ENODEV;
-- 
2.48.1

